# <center> The top brands amongst used cars on Ebay for Germans <center>

 
## <center> Introduction <center>

#### The [dataset](https://data.world/data-society/used-cars-data) was pulled by user "orgeleska". Said set focused on used car data from the Ebay site, targeted towards Germans. The objective is to identify the overall registration year of the cars, the time during which these cars were sold, the most frequent brands, as well as their price and mileage. Within the most frequent brands, it would be best to focus on the ones who make up the most of the used car pool.
 

## <center> Summary of Results <center>

#### The cars were sold from 2015 to April 2016. The selling period spiked mostly during March 2016 and peaked at its highest during April 6th, then the 7th and 5th of the same year. In terms of the registration year, 58% of used cars are from the 2000s, with the remainder dating mostly to 1999 and older. As for the car brands, the most recurring brands were "volkswagen" (21% of the pool), "bmw", "mercedes_benz", "opel", "audi", "ford". These top 6 brands totaled over 60% of the used car pool. Lastly, the car's mileage was moderately high as well as their mean registration year, 12 to 14 years old. However, their prices were much more telling about the behavior of the top most frequent brands. The low end was occupied by the "opel"at 3,394. The high end was near the 10,000 euro mark with the "audi", "bmw", and "mercedez_benz", while the much more frequent "volkswagen" was at a mean of 6,000 euros. The prices were very spread out, but generally expensive.

## <center> Process <center>

### 1. Reading and learning the file

In [27]:
import pandas as pd

autos = (pd.read_csv('autos.csv',  
                    encoding = 'Latin-1')
                    )#Didn't run with UTF-8



In [28]:
#For a quick view of the records

autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [29]:
#Overview of file itself

autos.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [30]:
#Finding the minimum and maximum null data loss

print("The minimum loss is " + str(round(47320/50000*100)) + '.')
print('\n')
print("The maximum loss is " + str(round(40171/50000*100)) +'.')


The minimum loss is 95.


The maximum loss is 80.


#### There is an overall total of 50,000 rows and 20 columns. Said columns' datatypes are subdivided into 5 of int64 and 15 of object. However, some cleaning and conversions may be necessary. Depending on the needs, "datecrawled", "price", "odometer", "lastSeen", and "dateCreated" can be converted to numeric fields, for a total of five columns. In addition, rows had null fields in the columns of "vehicleType", "gearbox", "model", "fuelType" and "notRepairedDamage". Said null fields present a data loss from roughly 5% to a maximum of 20%.

### 2. Cleaning the column names

In [31]:
#Viewing the values to fix

autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [32]:
col_name_corrections = [
    'date_crawled', 
    'name', 
    'seller', 
    'offer_type',
    'price',
    'abtest',
    'vehicle_type', 
    'registration_year',
    'gearbox', 
    'powerPS',
    'model',
    'odometer', 
    'registration_month',
    'fuel_type',
    'brand',
    'unrepaired_damage', 
    'ad_created', 
    'nr_of_pictures',
    'postal_code',
    'last_seen']

autos.columns = col_name_corrections

#To view and validate corrections
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

#### The reason for changing from camelcase to snakecase was to facilitate the readability of the column names. 

In [33]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

### 3. Cleaning the essential data

#### This section focuses on identifying issues with the columns of data. Some of them will be addressed as they pertain to the needs of the analysis while the others are for anyone who may take interest in working with said data.

### Identifying issues with the numbers

In [34]:
#This has a decent amount of improbable car prices
print(autos["price"]
 .value_counts(dropna=False, ascending = False)
 .head()
)
autos["price"].dtype

$0        1421
$500       781
$1,500     734
$2,500     643
$1,000     639
Name: price, dtype: int64


dtype('O')

In [35]:
#Getting a sense for the format of the data
print(autos["odometer"]
 .value_counts(dropna=False)
 .head()
)
print('\n','Unique Values : ',
      autos["odometer"].unique(),'\n')

autos["odometer"].dtype

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
Name: odometer, dtype: int64

 Unique Values :  ['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km'] 



dtype('O')

#### The first inconsistency to address is the "object" type that price and odometer have. The second inconsistency is the appearance of strange values in "price". It is very unlikely for a used car to retail for 0 dollars, even if it was a broken car exclusively meant for harvesting parts. . In contrast, odometer simply needs to be formatted appropriately for further analysis with the removal of the "km" and the commas.

### Cleaning the numbers 

In [36]:
#Fixing the price column's data and type

autos['price'] = autos['price'].str.replace('$','')

autos['price'] = autos['price'].str.replace(',','')

autos['price'] = autos['price'].astype(int)


#To check results
print(autos["price"]
 .value_counts(dropna=False, ascending = False)
 .head()
)
autos["price"].dtype

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64


dtype('int64')

In [37]:
#Fixing the odometer column data and type

autos['odometer'] = autos['odometer'].str.replace(',','')

autos['odometer'] = autos['odometer'].str.replace('km','')

autos['odometer'] = autos['odometer'].astype(int)

autos.rename({'odometer':'odometer_km'}, axis = 'columns', inplace = True)

#To check results

print(autos['odometer_km']
     .value_counts(dropna = False)
     .head()
     )

autos['odometer_km'].dtype

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64


dtype('int64')

### Analyzing price and odometer km

In [38]:
#Reviewing the data in the 'price' column

print('Number of unique values :', autos['price'].unique()
      .shape[0]#Counting unique entries
     , '\n')

print(autos['price'].describe(),'\n')#Now, it is int64.

print(autos['price'].value_counts()
              .sort_index(ascending = False)
              .head(10)#To find highest outliers
               )
autos['price'].between(0,499).value_counts()
#To gauge amount of values between 0 & 499 

Number of unique values : 2357 

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64 

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64


False    45111
True      4889
Name: price, dtype: int64

#### Upon careful consideration and observation, values above 1 million and below 500 will be treated as outliers and  irrelevant to the analysis. The reasoning for eliminating the values below 500 is that they seem too low and unrealistic, such as the 1,421 of used cars selling for a price of 0. To study the outliers above 1 million, the following code block was run where a custom sorted 'autos' dataframe with the registration year revealed oddities. Among them, there were either extremely high values such as a 99.9 million euro price for a 1992 Ferrari or strange values, almost akin to a jest, such as "12,345,678", "11,111,111", "1,234,566". For the latter, prices do not tend to reiterate the same number or run through all the digits from "0" to "9", in contrast to the retail patterns of having any type of price end in the digits ".99" or "9.99". Lastly, the descriptive statistics will be run again with these conditions in place. 

In [39]:
(autos[['price','registration_year','name']
      ].sort_values(by = ['price'], ascending= False)
       .head(15)
      )

price  registration_year  \
39705  99999999               1999   
42221  27322222               2014   
27371  12345678               2017   
39377  12345678               2018   
47598  12345678               2001   
2897   11111111               1973   
24384  11111111               1995   
11137  10000000               1960   
47634   3890000               2006   
7814    1300000               1992   
22947   1234566               1999   
43049    999999               1981   
514      999999               2009   
37585    999990               1985   
36818    350000               2016   

                                                    name  
39705                        Tausch_gegen_gleichwertiges  
42221                                  Leasinguebernahme  
27371                                         Fiat_Punto  
39377                        Tausche_volvo_v40_gegen_van  
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...  
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000  
24384                            Schlachte_Golf_3_gt_tdi  
11137  suche_maserati_3200_gt_Zustand_unwichtig_laufe...  
47634                                        Ferrari_FXX  
7814                                         Ferrari_F40  
22947                         Bmw_530d_zum_ausschlachten  
43049                                      2_VW_Busse_T3  
514                     Ford_Focus_Turnier_1.6_16V_Style  
37585                                Volkswagen_Jetta_GT  
36818                                        Porsche_991

In [40]:
#Removing the outliers

clean_price = autos['price'].between(500,1000000)

autos = autos[clean_price] 

#Running the code again under more relevant parameters

               
autos['price'].describe()
#With new conditions, the findings are less obscured by 
#outliers.


count     45100.000000
mean       6386.757738
std       12306.631453
min         500.000000
25%        1500.000000
50%        3500.000000
75%        7900.000000
max      999999.000000
Name: price, dtype: float64

####  Despite a 6,386 average, most cars were selling near the 3,500 dollar mark and in an overall range of 1,500 to 7,900. However, the standard deviation is a bit high given the wide price range for these cars. Nevertheless, this serves as a preliminary view of the bulk of the prices for these used cars. Now, "odemeter_km" follows.

In [41]:
#Reviewing the data in the 'odometer' column

print('Number of unique values :', autos['odometer_km'].unique()
      .shape[0]#Counting unique entries
     , '\n')

print(autos['odometer_km'].describe(),'\n')#Now, it is int64.

print(autos['odometer_km'].value_counts()
              .sort_index(ascending = False)
              .head(5)#To find potential outliers
               )


Number of unique values : 13 

count     45100.000000
mean     125294.124169
std       39621.768672
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64 

150000    28700
125000     4839
100000     2031
90000      1676
80000      1385
Name: odometer_km, dtype: int64


####  The bulk of the data is in the 150,000 km area, validated by its 50th and 75th percentile. 28,700 of 45,100 records of cars show these vehicles running this amount of distance, which is quite great. In other words, these cars have definitely seen quite a bit of use.

### Looking into the dates

In [43]:
#Showing the oddities with the hours

print('First few rows:', '\n',autos["date_crawled"]
      .head()
     ) #To get a sense of the data

print('\n','Empty entries :', autos["date_crawled"]
      .isnull()
      .value_counts(),
      '\n'
     )# To find number of blanks specifically

print('dtype : ',autos["date_crawled"].dtype, '\n')
#To get confirmation on the type


#Cleaning the column's data

autos["date_crawled"] = autos["date_crawled"].str[:10]


#To get a sense of the distribution in dates

print('Distribution :','\n',
      autos["date_crawled"].value_counts
      (normalize = True,dropna = False)
     )


First few rows: 
 0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object

 Empty entries : False    45100
Name: date_crawled, dtype: int64 

dtype :  object 

Distribution : 
 2016-04-03    0.038825
2016-03-20    0.038071
2016-03-21    0.037761
2016-03-12    0.037317
2016-04-04    0.036608
2016-03-14    0.036297
2016-03-07    0.036186
2016-04-02    0.035765
2016-03-28    0.034834
2016-03-19    0.034745
2016-03-15    0.034013
2016-04-01    0.033902
2016-03-30    0.033326
2016-03-29    0.033282
2016-03-08    0.033171
2016-03-11    0.033016
2016-03-22    0.033016
2016-03-09    0.032905
2016-03-10    0.032705
2016-03-26    0.032639
2016-03-23    0.032395
2016-03-31    0.031663
2016-03-17    0.031175
2016-03-27    0.031175
2016-03-25    0.031086
2016-03-16    0.029357
2016-03-24    0.028980
2016-03-05    0.025565
2016-03-13    0.015521
2016-03-06    0.014124
2016-04-05    0.013171
2016-03-18    0.012882
2016-04-06    0.003171
2016-0

#### The 'date_crawled' field has no empty entries, is an object type, and, after cleaning, has a range of dates from march to early april 2016 with a high frequency in '2016-04-03'. This last finding is interesting when the bulk of the data is concentrated in march. However, this field is the date of the data's retrieval. The next is 'ad_created'.

In [44]:
#Same timeslot, different date

print('Showing timeslot pattern :', '\n',
      autos["ad_created"]
      .value_counts(dropna=False)
      .head()
     )#To show the common timeslot amongst records

print('\n','Empty Entries : ', autos["ad_created"]
      .isnull()
      .value_counts(),
      '\n'
     )# To find number of blanks specifically


print('dtype : ', autos["ad_created"].dtype,'\n')
#To get confirmation on the type


#Cleaning the column's data

autos["ad_created"] = autos["ad_created"].str[:10]


#To get a sense of the distribution in dates

print('The highest frequencies: ','\n',
      autos["ad_created"].value_counts
      (normalize = True, 
       dropna = False)
      .head(30), '\n'# For the highest value
     )

print(autos["ad_created"].value_counts
      (normalize = True, 
       dropna = False)
      .sort_index()
      .head(3), '\n'# To know the oldest date
     )

print(autos["ad_created"].value_counts
      (normalize = True, 
       dropna = False)
      .sort_index(ascending = False)
      .head(3)# To know the latest date
     )


Showing timeslot pattern : 
 2016-04-03 00:00:00    1761
2016-03-20 00:00:00    1723
2016-03-21 00:00:00    1713
2016-03-12 00:00:00    1673
2016-04-04 00:00:00    1668
Name: ad_created, dtype: int64

 Empty Entries :  False    45100
Name: ad_created, dtype: int64 

dtype :  object 

The highest frequencies:  
 2016-04-03    0.039047
2016-03-20    0.038204
2016-03-21    0.037982
2016-03-12    0.037095
2016-04-04    0.036984
2016-04-02    0.035455
2016-03-07    0.034900
2016-03-14    0.034878
2016-03-28    0.034878
2016-04-01    0.033836
2016-03-15    0.033792
2016-03-19    0.033614
2016-03-11    0.033326
2016-03-29    0.033259
2016-03-30    0.033193
2016-03-08    0.033126
2016-03-09    0.032993
2016-03-22    0.032772
2016-03-26    0.032639
2016-03-10    0.032439
2016-03-23    0.032217
2016-03-31    0.031729
2016-03-25    0.031242
2016-03-27    0.031131
2016-03-17    0.030843
2016-03-16    0.029845
2016-03-24    0.028958
2016-03-05    0.023126
2016-03-13    0.016962
2016-03-06    0.0153

####  The "ad_created" column contains a few interesting findings. Most of the ads were created in April 3rd, 2016. On the other hand, that is essentially a max value of 3.9% of all ads. The bulk of the ads were created in March, mostly towards the end and beginning of the month. As for the oldest ad on record, one from June 11th, 2015 was found. Lastly, there were no empty entries and the datatype was object. Next is the "last_seen" column".

In [45]:
#Now for "last_seen"
print('Observing a few records :', '\n',
      autos["last_seen"]
      .value_counts(dropna=False)
      .head()
     )#To show the common timeslot amongst records

print('\n','Empty Entries : ', autos["last_seen"]
      .isnull()
      .value_counts(),
      '\n'
     )# To find number of blanks specifically


print('dtype : ', autos["last_seen"].dtype,'\n')
#To get confirmation on the type


#Cleaning the column's data

autos["last_seen"] = autos["last_seen"].str[:10]


#To get a sense of the distribution in dates

print('The highest frequencies: ','\n',
      autos["last_seen"].value_counts
      (normalize = True, 
       dropna = False)
      .head(5), '\n'# For the highest value
     )


Observing a few records : 
 2016-04-07 06:17:27    8
2016-04-07 03:16:17    7
2016-04-06 21:17:51    7
2016-04-06 20:48:27    6
2016-04-06 14:17:04    6
Name: last_seen, dtype: int64

 Empty Entries :  False    45100
Name: last_seen, dtype: int64 

dtype :  object 

The highest frequencies:  
 2016-04-06    0.225322
2016-04-07    0.134146
2016-04-05    0.126608
2016-03-17    0.027672
2016-04-03    0.024945
Name: last_seen, dtype: float64 



#### Similar to "date_crawled", "last_seen" was created during the retrieval of the data. In similar fashion to the previous fields, there were no empty entries and the datatype was object. As for the findings, the bulk of the ads themselves closed (or were last seen) during the beginning of April 2016, mostly the 6th, 7th and 5th. This coincides with the majority of the ads being created in march. It can be summarized that the timing of car sales tend to peak in March. Such high activity could potentially indicate the buying of new car models and the selling of old ones.  

In [46]:
#This column has much older and newer years.
print('There are strange years in this column : ','\n',
      autos["registration_year"]
      .value_counts(dropna=False, ascending = True)
      .sort_index()#To see the oldest unlikely years
      .head(10), '\n'
     )

print(autos["registration_year"].describe(),'\n','\n',
     '1000 is clearly too old and 9999 is beyond 2016.'
     )

print('\n','Empty entries : ',
      autos["registration_year"]
      .isnull()
      .value_counts(),
      '\n'
     )

print('dtype : ', autos["registration_year"].dtype, '\n')

#Finding cutoff years for the "registration_year" column


print("Total number of cars with registration years before 1927 and after 2016 : ",
autos[(autos['registration_year'] < 1927) 
      | 
      (autos['registration_year'] > 2016)].shape[0])
    #The earliest feasible year is 1927.
    #The latest year is 2016.

There are strange years in this column :  
 1000    1
1001    1
1910    2
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
Name: registration_year, dtype: int64 

count    45100.000000
mean      2005.063282
std         89.649158
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64 
 
 1000 is clearly too old and 9999 is beyond 2016.

 Empty entries :  False    45100
Name: registration_year, dtype: int64 

dtype :  int64 

Total number of cars with registration years before 1927 and after 2016 :  1776


#### Cars before 1927 and after 2016 will be culled from the data. As for 2016, it is the year corresponding to the latest ad created. The reason for choosing 1927 is the accuracy of the registration year in reference to a car's release date. In other words, the earliest feasible car records occur with a registration year of 1927. On the other hand, 1000 and 1001 aren't viable, which leaves 1910. Upon  running the following code to isolate records with that registration year, the car's information was searched and verified on the internet. Both of those cars hail from the 1990s, 1994 and 1992 respectively. In conclusion, these new cutoff years will allow for clearer results for the "registration_year" column, and a loss of 1,776 records is quite acceptable as a result of his cutoff. 

In [47]:
(autos[autos['registration_year'] == 1910])

date_crawled            name  seller offer_type  price   abtest  \
22659   2016-03-14    Opel_Corsa_B  privat    Angebot    500     test   
28693   2016-03-22  Renault_Twingo  privat    Angebot    599  control   

      vehicle_type  registration_year  gearbox  powerPS  model  odometer_km  \
22659          NaN               1910      NaN        0  corsa       150000   
28693   kleinwagen               1910  manuell        0    NaN         5000   

       registration_month fuel_type    brand unrepaired_damage  ad_created  \
22659                   0       NaN     opel               NaN  2016-03-14   
28693                   0    benzin  renault               NaN  2016-03-22   

       nr_of_pictures  postal_code   last_seen  
22659               0        52393  2016-04-03  
28693               0        70376  2016-04-06

In [48]:
(autos[autos['registration_year'] == 1927])

date_crawled                     name  seller offer_type  price  \
21416   2016-03-12  Essex_super_six__Ford_A  privat    Angebot  16500   

        abtest vehicle_type  registration_year  gearbox  powerPS   model  \
21416  control       cabrio               1927  manuell       40  andere   

       odometer_km  registration_month fuel_type brand unrepaired_damage  \
21416         5000                   5    benzin  ford               NaN   

       ad_created  nr_of_pictures  postal_code   last_seen  
21416  2016-03-12               0        74821  2016-03-15

In [49]:
cutoff_bool = (autos['registration_year']
               .between(1927,2016)
              )

autos = autos[cutoff_bool]

print(autos["registration_year"].describe())

In_the_2000s = autos["registration_year"].between(2000,2009)

print('\n',"% of cars from the 2000s : ",
      (autos[In_the_2000s].shape[0])/autos.shape[0]
     )


count    43324.000000
mean      2003.230773
std          7.048142
min       1927.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

 % of cars from the 2000s :  0.5788708337180316


#### This is clearer and more consistent. Without extremely high and low values distorting the mean, the result is 2003, which closely matches the overall median of 2004. In other words, most used cars sold here are from the 2000s (58%) with some older and newer cars as well. 

### Aggregating by top 20 brands

In [50]:
top_20_brands = (autos["brand"]
          .value_counts(normalize = True, dropna = False)
          .head(20)
         ) #The most frequent brands of the used car pool

print(top_20_brands)

top_20_aggregate = {} #First dictionary for mean prices


#To pair alphabetically with numbers

letter_rank_holder = 'ABCDEFGHIJKLMNOPQRST' #Only 20


#For maintaining the consistency of the ranking

rank_tally = 1 #The actual ranking (1st, 2nd, 3rd)

letter_tally = rank_tally - 1 #To preserve the pairing


for brand in top_20_brands.index: 
    
    top_20_aggregate[letter_rank_holder[letter_tally]
                      + str(rank_tally)
                      + '. '
                      + str(brand)] = brand
    
    rank_tally += 1 
    
    letter_tally += 1 #The pairings are A1, B2, C3, D4,
                      #and so forth.
        
    #The dictionary will be something akin to 
    #{A1. Name of most frequent brand of used car :
    #Name of most frequent brand of used car,
    #B2. Name of second most frequent brand of used car:
    #Name of second most frequent brand of used car}
    
    #Pairings like A1 or C3 are meant to circumvent
    #the sorted(dictionary) sorting numbers oddly,
    #such as a pattern of "1. Brand A", "10. Brand X", 
    #and then "2. Brand D".
    

top_20_sorting_copy = top_20_aggregate.copy()
#This is for a bit later. The copy will be used for
#dataframe sorting. For now, the focus is a quick
#preliminary view.

for rank in top_20_aggregate:
    
    brand = top_20_aggregate[rank] #The unaltered 
                                   #brand name
    
    price_mean =(
        autos.loc[autos['brand'] == brand,'price']
        .mean()
        )
    
    price_mean = int(price_mean)
    
    top_20_aggregate[rank] = price_mean 
    
    #Changing the value now to the price_mean, now
    #that the unaltered brand name (the one without the
    #pairings) has served its purpose for the boolean.
    

print('\n', 'Mean prices of the top 20 brands : ', ' \n')

for brand in sorted(top_20_aggregate):
    
    print(brand, ' : ', top_20_aggregate[brand])

    


volkswagen        0.211938
bmw               0.115017
mercedes_benz     0.102253
opel              0.100314
audi              0.090504
ford              0.065183
renault           0.043879
peugeot           0.029360
fiat              0.023520
seat              0.017888
skoda             0.017150
smart             0.015119
nissan            0.015003
mazda             0.015003
citroen           0.014149
toyota            0.013411
hyundai           0.010156
sonstige_autos    0.009694
volvo             0.009371
mini              0.009371
Name: brand, dtype: float64

 Mean prices of the top 20 brands :   

A1. volkswagen  :  6000
B2. bmw  :  8582
C3. mercedes_benz  :  8766
D4. opel  :  3394
E5. audi  :  9613
F6. ford  :  4644
G7. renault  :  2820
H8. peugeot  :  3360
I9. fiat  :  3256
J10. seat  :  4810
K11. skoda  :  6558
L12. smart  :  3611
M13. nissan  :  5175
N14. mazda  :  4459
O15. citroen  :  4012
P16. toyota  :  5266
Q17. hyundai  :  5686
R18. sonstige_autos  :  13439
S19. volvo  : 

#### After trying various workarounds, it was thought best to preserve the descending order of the most frequent car brands and simply adding the mean prices to the brands. This was done by using letters for maintaining set order and then numbers for readability. What is being observed is essentially the mean prices for the most frequent brands of used cars sold on ebay during the given timeframe. The "volkswagen", "bmw", "mercedes_benz", "opel",  "audi", and "ford" make up over 60% of the used car pool, where "volkswagen" comprises 21% of said pool while holding a lead of 10% over its nearest competitors. Other notable findings  were the expensive "sonstige_autos" (13,439) and mini (10,691) as brands with low frequency while the similarly priced audi (9,613) was the fifth most frequent brand. In addition, the "bmw", "mercedez_benz", and "audi" brands were typically priced for more than 8,000 euros except the "opel"  (3,394) and the "volkswagen" (6,000). The most frequent brands exhibit interesting behavior in the sense that the prices scale from affordable to very expensive. As for the less frequent vehicles, they were typically sold below or slightly above the 5,000 euro mark, where the cheapest car was the "renault" brand.

#### Storing the aggregate data for the top 6 most frequent brands (Profiling the top 6 brands)

#### The purpose of the previous block of code was to get a quick view at the most frequent brands of used cars and their mean prices. After the view, it was decided to focus on adding the mean mileage, preserve the sorting of the most frequent brands (instead of sorting by another metric), and to isolate the top 6 most frequent brands as they comprise most of the pool.

In [51]:
#The goal is to create a series from this dictionary, 
#while the past aggregate was for preliminary view.

top_20_mean_price = {} #Dictionary for mean prices


#The following dictionaries are for capturing mileages and
#years, as they will also become series in a custom 
#dataframe for the top 20 most frequent car brands.

top_20_mean_mileage = {}

top_20_registration_year = {}

#This last piece of the puzzle is to help  
#preserve sorting

frequency_rankings = {}

for item in top_20_sorting_copy: #The copy made before
    
    brand = top_20_sorting_copy[item] #The key
    
    rank = '' #The value
    
    for char in item:
        
        if char in '0123456789': 
            
            rank += char #To get numbers and no letters
    
    frequency_rankings[brand] = int(rank)
    
    #This dictionary will be used to make a series
    #that preserves the rankings of most frequent brands
    #exclusively for sorting purposes.

    
for brand in top_20_brands.index: 
    
    
    top_20_mean_price[brand] = 0
    
    top_20_mean_mileage[brand] = 0
    
    top_20_registration_year[brand] = 0
    
    
for brand in top_20_mean_price:
    
    price_mean =int((
        autos.loc[autos['brand'] == brand,'price']
        .mean()
        ))
    
    mean_mileage = int((
        autos.loc[autos['brand'] == brand,'odometer_km']
        .mean()
        ))
    
    registration_year_mean = int((
        autos.loc[autos['brand'] == brand,
        'registration_year']
        .mean()
        ))

    
    top_20_mean_price[brand] = price_mean 
    
    top_20_mean_mileage[brand] = mean_mileage
    
    top_20_registration_year[brand] = registration_year_mean
    

#Turning these dictionaries into a series

top_20_mean_price_series = pd.Series(top_20_mean_price)

top_20_mean_mileage_series = (pd.Series
                              (top_20_mean_mileage))

top_20_registration_year_series = (pd.Series
                              (top_20_registration_year))

frequency_rankings_series = pd.Series(frequency_rankings)

frequency_series = pd.Series(top_20_brands) 
#The frequencies themselves, not the rankings

top_20_brand_profile = (
    pd.DataFrame(top_20_mean_price_series,
    columns = ['mean_price'])
    )
                       
top_20_brand_profile['mean_mileage'] = (
    top_20_mean_mileage_series
    )

top_20_brand_profile['registration_year'] = (
    top_20_registration_year_series
    )

top_20_brand_profile['frequency_rankings'] = (
    frequency_rankings_series
    )

top_20_brand_profile['distribution'] = (
    frequency_series
    )

#Isolating the 6 most frequent brands and sorting them
top_6 = (
    top_20_brand_profile.sort_values
    (by = ['frequency_rankings']).head(6)
    )

top_6



mean_price  mean_mileage  registration_year  \
volkswagen           6000        128239               2002   
bmw                  8582        132865               2003   
mercedes_benz        8766        131019               2002   
opel                 3394        128007               2002   
audi                 9613        128909               2004   
ford                 4644        123495               2003   

               frequency_rankings  distribution  
volkswagen                      1      0.211938  
bmw                             2      0.115017  
mercedes_benz                   3      0.102253  
opel                            4      0.100314  
audi                            5      0.090504  
ford                            6      0.065183

#### The top 6 brands are very close to each other in mean mileage. The standouts are 'mercedes_benz' and 'bmw' for breaking the 130,000 km mark and 'ford' for being the lowest with 123,495 km. What is most interesting is that the mean mileage has no meaningful relationship on the price because pricey and cheaper cars have high odometer values. Nevertheless, the mileage was a way to gauge some of that 'new-ness' any product can retain if they are not used too much, which was definitely not the case here. Their mean age validates this, as they are potentially anywhere from 12 to 14 years old.

### For future reference: additional notes on potential cleaning of columns

#### If the data were to be cleaned further for other types of analyses, here are some patterns identified in some columns to facilitate the process.

### The One-Value Columns

In [52]:
#These are the columns with predominantly one value.
print(autos["seller"]
 .value_counts(dropna=False)
 .head()
)

autos["seller"].dtype

privat    43324
Name: seller, dtype: int64


dtype('O')

In [53]:
print(autos["offer_type"]
 .value_counts(dropna=False)
 .head()
)

autos["offer_type"].dtype

Angebot    43324
Name: offer_type, dtype: int64


dtype('O')

In [54]:
#No number of pictures indicates loss of data.
print(autos["nr_of_pictures"]
 .value_counts(dropna=False)
 .head()
)

autos["nr_of_pictures"].dtype

0    43324
Name: nr_of_pictures, dtype: int64


dtype('int64')

#### These columns, "seller" and "offer type" in particular, have predominantly one value and can be removed from the dataframe. Similarly, "nr_of_pictures" essentially has no data whatsoever and should also be removed. 

### Columns with String Data

In [55]:
#Verifying the rest to see if they are null or equal 0


remaining_col = ['unrepaired_damage',
                 'postal_code', 'name', 'vehicle_type',
                 'gearbox', 'model', 'fuel_type', 'brand']

print('NaNs in:','\n')

for col_name in remaining_col:
    
    print(col_name, ' : ',autos[col_name].isna().sum())

print('\n')

print('Columns with 0 as data :', '\n')

for col_name in remaining_col: 
    
    #Split these into two loops for 
    #readability of results
    
    
    print(col_name,' : ',
          (autos.loc[autos[col_name] == 0,col_name])
          .count())
                    

    
#The reasoning for using 0 is because some data, when 
#missing tends to be left empty or filled with a 0.
#Theoretically, text values should not be 0, which
#is something that can be validated.

NaNs in: 

unrepaired_damage  :  6766
postal_code  :  0
name  :  0
vehicle_type  :  1902
gearbox  :  1603
model  :  1817
fuel_type  :  2529
brand  :  0


Columns with 0 as data : 

unrepaired_damage  :  0
postal_code  :  0
name  :  0
vehicle_type  :  0
gearbox  :  0
model  :  0
fuel_type  :  0
brand  :  0


#### There are numerous instances of no data in "unrepaired_damage", "vehicle_type", "gearbox", "model" and "fuel_type". Depending on future analysis needs, they may need to be cleaned. Columns with a predominant number of strings that have not been singled out have no issue worth mentioning.

### Registration month and PowerPS

In [56]:
#This column has a zero month.
print(autos["registration_month"]
 .value_counts(dropna=False)
 .head()
)

print('\n',autos["registration_month"]
      .isnull()
      .value_counts(),
      '\n'
     )

autos["registration_month"].dtype

3    4543
6    3901
4    3671
5    3669
7    3531
Name: registration_month, dtype: int64

 False    43324
Name: registration_month, dtype: int64 



dtype('int64')

#### Although there are no blank records,  "registration month" has 5,075 ocurrences of a month of "0".  This could be an impedance to any month based analysis.

In [57]:
#A powerPS of 0 could be unlikely.
print(autos["powerPS"]
 .value_counts(dropna=False)
 .head()
)

autos["powerPS"].dtype

0      3506
75     2609
150    1874
140    1795
60     1637
Name: powerPS, dtype: int64


dtype('int64')

#### It is potentially unlikely that a working car has a horsepower of 0, so this may present hurdles to any analysis that may include this column. 

## <center> Conclusion <center>


#### The objectives were to identify the timeframe for the cars sold, their overall registration year distribution,  the most frequent brands, and isolate the top 6 of these with their price and mileage. The cars were sold from 2015 to April 2016, the date in which the data retrieval ended. During this timeframe, the cars were sold at a higher frequency during March 2016 and particularly peaked at April 6th, then the 7th and 5th of the same year. The overall registration year for the cars indicates that 25% of the used cars are equal to or older than 1999, while 58% are from the 2000s, and the remaining vehicles are from 2010 to 2016. As for the car brands, the most recurring brands were "volkswagen" (the highest), "bmw", "mercedes_benz", "opel", "audi", "ford". These top 6 brands totaled over 60% of the used car pool. Lastly, the car's mileage was moderately high, and the mean age ranged from 12 to 14 years old, but the prices were much more telling about the behavior of the top brands. The low end was occupied by the "opel"at 3,394. The high end was near the 10,000 euro mark with the "audi", "bmw", and "mercedez_benz", while the much more frequent "volkswagen" at a mean of 6,000 euros. Despite having cheaper and moderately expensive cars, the overall pricing was expensive. This is especially more compelling when one considers speculating on the original pricing of these cars. If they were selling at a theoretical 30,000 euros each, this would potentially mean used cars of said brands would be selling from 10% to 20% of their original price on the low end and 33%. Lastly, said speculation isn't even accounting for the effects of inflation over time, so these findings could lead to further analytic inquiries if necessary.
